docs/architecture/adr-036-blockchain-reactor-refactor.md

```diff
@@ -0,0 +1,406 @@
+* `Event` - significant events in the network.
+
+producing clear outputs:
+
+* `State` - updated state of the node,
+* `Message` - signal what message to send,
+* `TimeoutTrigger` - signal that timeout should be triggered.
+
+```go
+type Event int
+
+const (
+	EventUnknown Event = iota
+	EventStatusReport
+	EventBlockRequest
+	EventNewBlock
+	EventRemovePeer
+	EventResponseTimeout
+	EventTerminationTimeout
+)
+
+type BlockRequest struct {
+	Height int64
+	PeerID p2p.ID
+}
+
+type NewBlock struct {
+	Height        int64
+	Block         Block
+	Commit        Commit
+	PeerID        ID
+	CurrentHeight int64
+}
+
+type StatusReport struct {
+	PeerID ID
+	Height int64
+}
+
+type RemovePeer struct {
+	PeerID ID
+	Height int64
+}
+
+type ResponseTimeout struct {
+	PeerID ID
+	Height int64
+}
+
+type TerminationTimeout struct {
+	Height int64
+}
+
+type Message int
+
+const (
+	MessageUnknown Message = iota
+	MessageBlockRequest
+	MessageBlockResponse
+	MessageStatusReport
+)
+
+type BlockRequestMessage struct {
+	Height int64
+	PeerID ID
+}
+
+type BlockResponseMessage struct {
+	Height        int64
+	Block         Block
+	Commit        Commit
+	PeerID        ID
+	CurrentHeight int64
+}
+
+type StatusReportMessage struct {
+	Height int64
+}
+
+type TimeoutTrigger int
+
+const (
+	TimeoutUnknown TimeoutTrigger = iota
+	TimeoutResponseTrigger
+	TimeoutTerminationTrigger
+)
+```
+
+The Controller state machine can be in two modes (states): `FastSyncMode` when
+it is trying to catch up with the network by downloading committed blocks,
+and `ConsensusMode` in which it executes Tendermint consensus protocol. We
+assume that "fast sync" mode terminates once the Controller switch to
+`ConsensusMode`.  
+
+``` go
+type State int
+
+const (
+	StateUnknown State = iota
+	StateFastSyncMode
+	StateConsensusMode
+)
+
+type ControllerState struct {
+	Height             int64            // the first block that is not committed
+	State              State            // mode of operation of the state machine
+	PeerMap            map[ID]PeerStats // map of peer IDs to peer statistics
+	MaxRequestPending  int64            // maximum height of the pending requests
+	FailedRequests     []int64          // list of failed block requests
+	PendingRequestsNum int              // number of pending requests
+	Store              []BlockInfo      // contains list of downloaded blocks
+	Executor           BlockExecutor    // store, verify and executes blocks
+}
+
+type PeerStats struct {
+	Height         int64
+	PendingRequest int64 // it can be list in case there are multiple outstanding requests per peer
+}
+
+type BlockInfo struct {
+	Block  Block
+	Commit Commit
+	PeerID ID // a peer from which we downloading the corresponding Block and Commit
+}
+```
+
+The `Controller` is initialized by providing initial height (`startHeight`) from which it will start downloading blocks from peers.
+
+``` go
+func ControllerInit(state ControllerState, startHeight int64) ControllerState {
```

Comment:
Why does init get an exisiting state? Is it not supposed to create a fresh state for us?

If so a more idiomatic name would be `NewControllerState`.